src/crystal/event_loop/iocp.cr

```diff
@@ -1,146 +1,186 @@
-require "c/ioapiset"
-require "crystal/system/print_error"
+# forward declaration for the require below to not create a module
+class Crystal::EventLoop::IOCP < Crystal::EventLoop
+end
+
+require "c/ntdll"
 require "../system/win32/iocp"
+require "../system/win32/waitable_timer"
+require "./timers"
+require "./iocp/*"
 
 # :nodoc:
 class Crystal::EventLoop::IOCP < Crystal::EventLoop
-  # This is a list of resume and timeout events managed outside of IOCP.
-  @queue = Deque(Event).new
-
-  @lock = Crystal::SpinLock.new
-  @interrupted = Atomic(Bool).new(false)
-  @blocked_thread = Atomic(Thread?).new(nil)
+  @waitable_timer : System::WaitableTimer?
+  @timer_packet : LibC::HANDLE?
+  @timer_key : System::IOCP::CompletionKey?
+
+  def initialize
+    @mutex = Thread::Mutex.new
+    @timers = Timers(Timer).new
+
+    # the completion port
+    @iocp = System::IOCP.new
+
+    # custom completion to interrupt a blocking run
+    @interrupted = Atomic(Bool).new(false)
+    @interrupt_key = System::IOCP::CompletionKey.new(:interrupt)
+
+    # On Windows 10+ we leverage a high resolution timer with completion packet
+    # to notify a completion port; on legacy Windows we fallback to the low
+    # resolution timeout (~15.6ms)
+    @mutex.synchronize do
```

Comment:
Since `@mutex` was just created and the object was not initialized yet, does this have any effect?

Reply:
Oh right, it's local. It doesn't protect parallel calls to `wait_completion_packet_methods?` :facepalm:

Reply:
I tried a @@mutex on System::IOCP, but then I got into init var recursion failures... I didn't have any issues with `preview_mt` without a mutex, so either `LoadLibraryExW` is thread safe, or let's try our luck for now.